### PR TITLE
Ability to add tips

### PR DESCRIPTION
**Pull request by ryanmitchell**
Here we go! More to come in theme.

Ref theme changes: https://github.com/tastyigniter/ti-theme-orange/pull/5

### CODE REVIEW

**Comment by sampoyigi** [comment]
That was quick! Definitely a great start! 

I'm with @jeromio on supporting both fixed and percentage amounts. Also, multiple amounts to choose from like the [screenshot](https://github.com/tastyigniter/TastyIgniter/issues/256#issuecomment-626360447) by @justinbriansmith, its same on UberEats but with fixed amounts. 

Maybe add a repeater form field to the cart settings page to handle the amounts and a switch for fixed/percentage?

**Comment by ryanmitchell** [comment]
I've added the ability to toggle tipping on and off in cart settings, and also choose from some fixed percentages for now. 

It can handle both % and fixed values being passed to the field now, which covers both uses.

I will add the percentage value selection as a cart setting when I next get time to work on this.

**Comment by salhi197** [comment]
i would like to contribute in this if its possible

**Comment by ryanmitchell** [comment]
That’s great! Check out the pull requests I linked to and see if you can add anything.

**Comment by ryanmitchell** [comment]
@sampoyigi Check this over and let me know what you think. Between this and the theme changes it should be pretty much done.

**Comment by sampoyigi** [comment]
Ok, see the changes added, if you're okay with it I'll merge. Let me know if any clarifications are needed :)

**Comment by ryanmitchell** [comment]
Looks good, go for it.

**Comment by jeromio** [comment]
Question for @ryanmitchell: What provides the field_radiotoggle.php file? Seems like it would be in TastyIgniter itself, but I can't find a branch with that file?

**Comment by ryanmitchell** [comment]
It’s in the theme PR - the link is in the first comment.

**Comment by jeromio** [comment]
Nvmd - I found it. It's in TastyIgniter on the "develop" branch.

**Comment by jeromio** [comment]
Ok, next problem: 
Call to undefined method Igniter\Flame\Foundation\Http\Kernel::appendMiddlewareToGroup() on line 28 of Extension.php.

I tried figuring this out, but I'm stumped. This is not new code, but I never had this error before moving to the tipping branch. Any ideas?

**Comment by jeromio** [comment]
Maybe I can re-phrase this: How can I setup an environment (versions of each component) so that I can get TastyIgniter working with this new tipping feature?

**Comment by ryanmitchell** [comment]
@jeromio If you have the latest develop branch of TI, flame, local, cart and the orange theme it should be enough to get it working

I thing @sampoyigi is moving towards the next beta release of TI so if you are struggling to get it working it’s probably worth waiting for that.

**Comment by jeromio** [comment]
Ya, so I tried cloning each repo to get the latest from master or develop (not all components have a develop branch). That lead to lots of  hair pulling and basically just a borked site that didn't work at all.

SO I started over and discovered the 'edgeUpdates' => true switch in the docs. This is pretty cool and I was able to Update and find the new tipping feature in Cart settings. But the actual storefront site didn't work. I discovered that the edgeUpdates doesn't get develop branches where there were deps and it also seems to ignore the flame library code, leaving it too many versions behind. So I subbed in develop branch clones for flame and TI, but then there are weird deps on other versions of files - some of which don't appear to be in any available branch. An example is the Igniter\Flame\Foundation\Http\Kernel::appendMiddlewareToGroup() call in cart/Extension.php. That just doesn't exist anywhere? I commented that one out, then ran into several other discrepancies and it became a wack-a-mole of doom. I ended up deep into mealtimes and menu_item_options vs menu_options. Ironically (sadly?) I don't need that mealtimes feature. I really just need tipping - it's a critical feature. Not sure how to get the development config working to be able to run the latest code. Perhaps I just have bad luck (eg, I presume @ryanmitchell was able to run this and test it?).

Extra bummer bc I am now out of time and will have to go back to the other platform that I was using (weebly). It's kinda terrible, BUT, it does support tipping....

**Comment by ryanmitchell** [comment]
Sorry to hear that @jeromio 

@sampoyigi is currently packaging up and preparing a new beta release, though by the sounds of it it may be too late for you.

File: components/CartBox.php [hunk 1]
```diff
@@ -207,6 +210,32 @@ public function onApplyCoupon()
...
+    public function onApplyTip()
+    {
+        try {
+            $amountType = post('amount_type');
+            if (!in_array($amountType, ['none', 'amount', 'custom']))
+                throw new ApplicationException(lang('igniter.cart::default.alert_tip_not_applied'));
+
+            $amount = post('amount');
+//            if (preg_match('/^\d+([\.\d]{2})?([%])?$/', $amount) === FALSE)
+//                throw new ApplicationException(lang('igniter.cart::default.alert_tip_not_applied'));
```
Comment (sampoyigi) [inline components/CartBox.php:222]: uncomment?

File: cartconditions/Tip.php [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+        // if amount is not set, empty or 0
+        if (!$tipAmount = $this->getMetaData('amount'))
+            return FALSE;
+
+//        $value = $this->getMetaData('amount');
+//        if (preg_match('/^\d+([\.\d]{2})?([%])?$/', $value) === FALSE) {
+//            $this->removeMetaData('amount');
+//            flash()->warning(lang('igniter.cart::default.alert_tip_not_applied'))->now();
+//        }
```
Comment (sampoyigi) [inline cartconditions/Tip.php:40]: uncomment?

**Review by sampoyigi — COMMENTED** [review]
We missed these before merging 🤦

**Pull request merged**